### PR TITLE
Add e2e tests to the FSE plugin

### PR DESCRIPTION
**Pull request by noahtallen**
_Note: split from  #39000_

### Changes proposed in this Pull Request
* Uses @wordpress/scripts to run the e2e test commands.
* Run e2e tests against default environment URL
* Adds @wordpress/e2e-test-utils to help with e2e testing
* Adds an e2e test which tests to see if our back button override exists
* Adds a place for e2e helper functions, including one for activating a theme

### e2e tests
The e2e command automatically runs test files that:
1. Are in `apps/full-site-editing`
2. Have the extension `.spec.ts` or `.spec.js` OR is a js/ts file within a `specs` folder

### Testing instructions
1. Create and start a WordPress testing environment using #39000. (port must be localhost:4013)
2. Run `npm ci` from wp-calypso root
2. Run `npx lerna run build --scope='@automattic/full-site-editing' --stream --no-prefix` from wp-calypso root
2. Run `npx lerna run test:e2e --scope='@automattic/full-site-editing' --stream --no-prefix` from wp-calypso root
3. Tests should pass successfully

To run the e2e tests:
```sh
npx lerna run test:e2e --scope='@automattic/full-site-editing' --stream
```

### CODE REVIEW

**Comment by noahtallen** [comment]
I've temporarily based this on https://github.com/Automattic/wp-calypso/pull/39326, but once it is merged, I'll rebase on master. At that point, this should be good to go!

**Comment by vindl** [comment]
I'm currently running into the following error while trying to test this, but I think that my setup from #39000 might be broken:

```
> npx wp-scripts test-e2e --wordpress-base-url='http://localhost:4013'
(node:9610) UnhandledPromiseRejectionWarning: Error: Protocol error (Runtime.callFunctionOn): Target closed.
    at /Users/vindl/Automattic/wp-calypso/node_modules/puppeteer/lib/Connection.js:183:56
    at new Promise (<anonymous>)
    at CDPSession.send (/Users/vindl/Automattic/wp-calypso/node_modules/puppeteer/lib/Connection.js:182:12)
    at ExecutionContext._evaluateInternal (/Users/vindl/Automattic/wp-calypso/node_modules/puppeteer/lib/ExecutionContext.js:107:44)
    at ExecutionContext.evaluate (/Users/vindl/Automattic/wp-calypso/node_modules/puppeteer/lib/ExecutionContext.js:48:23)
    at ExecutionContext.<anonymous> (/Users/vindl/Automattic/wp-calypso/node_modules/puppeteer/lib/helper.js:112:23)
    at DOMWorld.evaluate (/Users/vindl/Automattic/wp-calypso/node_modules/puppeteer/lib/DOMWorld.js:112:20)
    at processTicksAndRejections (internal/process/task_queues.js:94:5)
  -- ASYNC --
    at Frame.<anonymous> (/Users/vindl/Automattic/wp-calypso/node_modules/puppeteer/lib/helper.js:111:15)
    at Page.evaluate (/Users/vindl/Automattic/wp-calypso/node_modules/puppeteer/lib/Page.js:863:43)
    at Page.evaluate (/Users/vindl/Automattic/wp-calypso/node_modules/puppeteer/lib/helper.js:112:23)
    at _callee$ (/Users/vindl/Automattic/wp-calypso/node_modules/@wordpress/e2e-test-utils/build/@wordpress/e2e-test-utils/src/create-new-post.js:32:42)
    at tryCatch (/Users/vindl/Automattic/wp-calypso/node_modules/regenerator-runtime/runtime.js:45:40)
    at Generator.invoke [as _invoke] (/Users/vindl/Automattic/wp-calypso/node_modules/regenerator-runtime/runtime.js:271:22)
    at Generator.prototype.<computed> [as next] (/Users/vindl/Automattic/wp-calypso/node_modules/regenerator-runtime/runtime.js:97:21)
    at asyncGeneratorStep (/Users/vindl/Automattic/wp-calypso/node_modules/@babel/runtime/helpers/asyncToGenerator.js:3:24)
    at _next (/Users/vindl/Automattic/wp-calypso/node_modules/@babel/runtime/helpers/asyncToGenerator.js:25:9)
    at processTicksAndRejections (internal/process/task_queues.js:94:5)
(node:9610) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:9610) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
FAIL full-site-editing-plugin/dotcom-fse/tests/fse-back-button.spec.js (13.266s)
  Full Site Editing Back Button
    ✕ Should have an overriden button (85ms)
  ● Full Site Editing Back Button › Should have an overriden button
    Timeout - Async callback was not invoked within the 10000ms timeout specified by jest.setTimeout.Error: Timeout - Async callback was not invoked within the 10000ms timeout specified by jest.setTimeout.
      at mapper (../../node_modules/jest-jasmine2/build/queueRunner.js:25:45)
  ● Full Site Editing Back Button › Should have an overriden button
    Execution context was destroyed, most likely because of a navigation.
      19 |      it( 'Should have an overriden button', async () => {
      20 |              // Target the toolbar__override class we added to see if our custom button exists.
    > 21 |              const button = await page.$(
         |                             ^
      22 |                      '.components-toolbar.edit-post-fullscreen-mode-close__toolbar.edit-post-fullscreen-mode-close__toolbar__override'
      23 |              );
      24 |              // Button is null if it does not exist.
      at rewriteError (../../node_modules/puppeteer/lib/ExecutionContext.js:167:15)
      at ExecutionContext._evaluateInternal (../../node_modules/puppeteer/lib/ExecutionContext.js:73:56)
      at ../../node_modules/puppeteer/lib/DOMWorld.js:132:24
      at DOMWorld.$ (../../node_modules/puppeteer/lib/DOMWorld.js:120:22)
      at Object.<anonymous> (full-site-editing-plugin/dotcom-fse/tests/fse-back-button.spec.js:21:18)
        -- ASYNC --
      at ExecutionContext.<anonymous> (../../node_modules/puppeteer/lib/helper.js:111:15)
      at ../../node_modules/puppeteer/lib/DOMWorld.js:132:38
        -- ASYNC --
      at Frame.<anonymous> (../../node_modules/puppeteer/lib/helper.js:111:15)
      at Page.$ (../../node_modules/puppeteer/lib/Page.js:319:29)
      at Page.$ (../../node_modules/puppeteer/lib/helper.js:112:23)
      at Object.<anonymous> (full-site-editing-plugin/dotcom-fse/tests/fse-back-button.spec.js:21:29)
Test Suites: 1 failed, 1 total
Tests:       1 failed, 1 total
Snapshots:   0 total
Time:        13.412s
Ran all test suites.
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! @automattic/full-site-editing@0.19.0 test:e2e: `npx wp-scripts test-e2e --wordpress-base-url='http://localhost:4013'`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the @automattic/full-site-editing@0.19.0 test:e2e script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/vindl/.npm/_logs/2020-02-11T23_41_11_340Z-debug.log
lerna ERR! npm run test:e2e exited 1 in '@automattic/full-site-editing'
```

**Comment by vindl** [comment]
Ok, I think I managed to fix my setup and I'm now seeing this:

```
> npx wp-scripts test-e2e --wordpress-base-url='http://localhost:4013'
FAIL full-site-editing-plugin/dotcom-fse/tests/fse-back-button.spec.js (5.506s)
  Full Site Editing Back Button
    ✕ Should have an overriden button (5ms)
  ● Full Site Editing Back Button › Should have an overriden button
    Evaluation failed: ReferenceError: wp is not defined
      at __puppeteer_evaluation_script__:2:15
      at ExecutionContext._evaluateInternal (../../node_modules/puppeteer/lib/ExecutionContext.js:122:13)
      at ExecutionContext.evaluate (../../node_modules/puppeteer/lib/ExecutionContext.js:48:12)
        -- ASYNC --
      at ExecutionContext.<anonymous> (../../node_modules/puppeteer/lib/helper.js:111:15)
      at DOMWorld.evaluate (../../node_modules/puppeteer/lib/DOMWorld.js:112:20)
        -- ASYNC --
      at Frame.<anonymous> (../../node_modules/puppeteer/lib/helper.js:111:15)
      at Page.evaluate (../../node_modules/puppeteer/lib/Page.js:863:43)
      at Page.evaluate (../../node_modules/puppeteer/lib/helper.js:112:23)
      at _callee$ (../../node_modules/@wordpress/e2e-test-utils/build/@wordpress/e2e-test-utils/src/create-new-post.js:32:42)
      at tryCatch (../../node_modules/regenerator-runtime/runtime.js:45:40)
      at Generator.invoke [as _invoke] (../../node_modules/regenerator-runtime/runtime.js:271:22)
      at Generator.prototype.<computed> [as next] (../../node_modules/regenerator-runtime/runtime.js:97:21)
      at asyncGeneratorStep (../../node_modules/@babel/runtime/helpers/asyncToGenerator.js:3:24)
      at _next (../../node_modules/@babel/runtime/helpers/asyncToGenerator.js:25:9)
  ● Full Site Editing Back Button › Should have an overriden button
    expect(received).toBeTruthy()
    Received: null
      23 |              );
      24 |              // Button is null if it does not exist.
    > 25 |              expect( button ).toBeTruthy();
         |                               ^
      26 |      } );
      27 | } );
      28 | 
      at Object.<anonymous> (full-site-editing-plugin/dotcom-fse/tests/fse-back-button.spec.js:25:20)
Test Suites: 1 failed, 1 total
Tests:       1 failed, 1 total
Snapshots:   0 total
Time:        5.579s, estimated 6s
Ran all test suites.
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! @automattic/full-site-editing@0.19.0 test:e2e: `npx wp-scripts test-e2e --wordpress-base-url='http://localhost:4013'`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the @automattic/full-site-editing@0.19.0 test:e2e script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/vindl/.npm/_logs/2020-02-11T23_50_04_962Z-debug.log
lerna ERR! npm run test:e2e exited 1 in '@automattic/full-site-editing'
```

File: .eslintrc.js [hunk 1]
```diff
@@ -33,6 +33,15 @@ module.exports = {
 				step: false,
 			},
 		},
+		{
```
Comment (tyxla) [inline .eslintrc.js]: Should this live in the `.eslintrc.js` of the FSE app?

File: package.json [hunk 2]
```diff
@@ -178,6 +178,7 @@
 		"@wordpress/babel-plugin-makepot": "3.3.0",
 		"@wordpress/browserslist-config": "2.6.0",
 		"@wordpress/dependency-extraction-webpack-plugin": "2.1.0",
+		"@wordpress/e2e-test-utils": "4.3.0",
```
Comment (tyxla) [inline package.json]: I wonder if this should live in the `devDependencies`  of the FSE app?

File: .eslintrc.js (hunk 1 above)
Comment (noahtallen) [reply .eslintrc.js]: Yeah, I think that would make more sense :)

File: package.json (hunk 2 above)
Comment (noahtallen) [reply package.json]: If I recall, dev dependencies are supposed to be specified in the root package.json (see p4TIVU-9l4-wp). But things are changing so quickly around dependency management in calypso that I'm not really sure :)

File: .eslintrc.js (hunk 1 above)
Comment (noahtallen) [reply .eslintrc.js]: Done!

File: package.json (hunk 2 above)
Comment (tyxla) [reply package.json]: I thought that might be the case, just wanted to confirm 👍

**Comment by vindl** [comment]
The previous errors were caused by because of the [old WP version in my Docker image](https://github.com/Automattic/wp-calypso/pull/39000#issuecomment-586039903).

**Review by vindl — APPROVED** [review]
LGTM! :shipit:

File: apps/full-site-editing/.eslintrc.js [hunk 3]
```diff
@@ -5,4 +5,12 @@ module.exports = {
 		'import/no-extraneous-dependencies': [ 'error', { packageDir: __dirname } ],
 		'react/react-in-jsx-scope': 0,
 	},
+	overrides: [
+		{
+			files: [ './**/?(*.)spec.[jt]s?(x)', './full-site-editing-plugin/e2e-test-helpers/**' ],
+			globals: {
```
Comment (vindl) [inline apps/full-site-editing/.eslintrc.js]: I'm curious about the purpose of this setting?

File: apps/full-site-editing/.eslintrc.js [hunk 4]
```diff
@@ -5,4 +5,12 @@ module.exports = {
...
+				page: 'readonly',
```
Comment (vindl) [inline apps/full-site-editing/.eslintrc.js]: I'm curious about the purpose of this setting?

Comment (noahtallen) [reply apps/full-site-editing/.eslintrc.js]: If we don't set this rule, then eslint throws an error when you try to access the `page` object from puppeteer. This tells eslint that the page global object is always available in the e2e tests (which it is)

**Comment by simison** [comment]
Needs a rebase :-)

File: package.json [hunk 5]
```diff
@@ -33,6 +33,7 @@
 		]
 	},
 	"dependencies": {
+		"@automattic/calypso-analytics": "file:packages/calypso-analytics",
```
Comment (noahtallen) [inline package.json]: I guess npm decided to move this for some reason

**Comment by noahtallen** [comment]
I haven't merged this yet because I'm still testing stuff with wp-env and wordpress/scripts to make sure it's working as expected.

File: package-lock.json [hunk 6]
```diff
@@ -6746,6 +6746,18 @@
 				"@babel/runtime": "^7.8.3"
 			}
 		},
+		"@wordpress/e2e-test-utils": {
```
Comment (noahtallen) [inline package-lock.json]: I think this is the first time I have ever seen a reasonable package.lock change

**Pull request merged**